source/features/add-pr-branches.js

```diff
@@ -0,0 +1,107 @@
+import {h} from 'dom-chef';
+import select from 'select-dom';
+import * as cache from '../libs/cache';
+import * as api from '../libs/api';
+import {getOwnerAndRepo} from '../libs/page-detect';
+
+function buildQuery(owner, repo, numbers) {
+	let query = `
+{
+  repository(owner: "${owner}", name: "${repo}") {`;
+
+	for (const number of numbers) {
+		query += `
+    ${number}: pullRequest(number: ${number.replace('issue_', '')}) {
+      baseRef {
+        name
+        repository {
+          url
+          owner {
+            login
+          }
+        }
+      }
+      headRef {
+        name
+        repository {
+          url
+          owner {
+            login
+          }
+        }
+      }
+    }`;
+	}
+
+	query += `
+  }
+}`;
+	return query;
+}
+
+async function fetchFromApi(owner, repo, numbers) {
+	const query = buildQuery(owner, repo, numbers);
+	const response = await api.v4(query);
+	console.log('add-pr-branches 3:', numbers, response);
+
+	if (response.data && response.data.repository) {
+		const d = {};
+		for (const [id, data] of Object.entries(response.data.repository)) {
+			d[id] = extractBranches(owner, data);
+		}
+		return d;
+	}
+}
+
+function extractBranches(owner, data) {
+	console.log('add-pr-branches 5:', data);
+
+	return {
+		base: {
+			label: data.baseRef.name,
+			url: `${data.baseRef.repository.url}/tree/${data.baseRef.name}`
+		},
+		head: {
+			label: (data.headRef.repository.owner.login === owner ? data.headRef.name : `${data.headRef.repository.owner.login}:${data.headRef.name}`),
+			url: `${data.headRef.repository.url}/tree/${data.headRef.name}`
+		}
+	};
+}
+
+function getPullBranches(owner, repo, number) {
+	return cache.getSet(`pull-branches:${owner}/${repo}`,
+		() => fetchFromApi(owner, repo, number)
```

Comment:
Sadly this is not that simple. The PR numbers on the current page are not the only ones that a repo has. If you want to use the cache you have to:

- cache each PR separately
- exclude cached PRs from the current view's fetch
- merge the previously-cached PRs with the fetch's result.

It would be nice, but as a workaround you can drop the caching altogether

Reply:
Yeah, I needed to rework this part when moving to the one query. I'll drop it, until the caching lib actually works (I couldn't get it to work).